Fix pen hiding

wentasah opened the pull request:
In versions before 872c905 ("Ignore the pointer_hidden flag in the drawing mode", 2020-05-09), it could happen, that a drawing mode tool was hidden and the user did not see where they will draw. The (now reverted) commit 872c905 fixed this problem by never hiding the drawing tool. This resulted in the drawing tool being visible in the slide area even if the cursor was moved out of the slide area.

The second commit in this PR fixes the problem differently. The drawing tool is hidden the same way as the pointer - after two seconds of inactivity or when the cursor is out of slide area. To make this work, it is sufficient to restart the timeout timer in the on_move_pen() callback.

fnevgeny commented:
Yes, I was aware that the drawing tool (pen/eraser) remains visible, but is it a problem? Contrary to the pointer proper, the pen is not visible on the *presentation* screen, so why bother?

wentasah commented:
Maybe it's not a problem, but it distracts me a bit during the presentations. Especially if the pen has bigger size.

fnevgeny commented:
Hmm, for me it doesn't switch off after the timeout.

wentasah commented:
Interesting. I tried my version again and also a new one rebased to current master and the pen (as well as the eraser) always hides after 2 seconds no matter which combination of action (moves, tool selection) I do. I'm testing it with `--no-install`, but I guess it should not matter.

My environment is Vala 0.48.9, gdk-pixbuf-2.40.0, cairo-1.16.0, graphite2-1.3.14, glib-2.66.2, gtk+3-3.24.23. Do you see anything what could cause the different behavior?

I noticed only a minor problem, that for the first time, both pointer and pen appear only after the first mouse move. I think it will be easy to fix.

fnevgeny commented:
Well, to make it work I need to add queue_pen_surface_draws() in restart_pointer_timer() - similar to queue_pointer_surface_draws(). If for you it works as is, something must be causing the (extraneous) widget redrawing(s). Window manager? I wonder if you experience the #553 issue, too?

wentasah commented:
I use i3 window manager. I'll try it with other WMs and try to debug the problem. For 553, I'll reply there.

wentasah commented:
I tried multiple WMs (i3, sway, gnome) and it seems that the type of WM is not important. It's important whether pdfpc runs under Wayland or X11. Under X11 (i3, gnome) the pen hides and I also experience high CPU load (as I wrote to #553). On Wayland (sway, gnome), the pen behaves differently depending on whether I use tablet or mouse. When using tablet, the pen (and also the pointer) hides immediately (not after 2 seconds), after raising the stylus. With mouse, the pen never hides. Also, if I use tablet, then mouse and then tablet again, then the pen stops hiding.

So it seems there are more bugs. I'll continue investigating this later.

wentasah commented:
I tried to investigate the different behavior between X11 and Wayland, but I no longer see any difference. I updated my system in the mean time, so this could be the reason. I also found that pen redraw is triggered by refresh of timer display, which happens every second (but not on the first slide). This could also be the reason why the pen was hiding for me.

Nevertheless, I added `queue_pen_surface_draws()` as @fnevgeny pointed out and rebased to current master.
